File: server/src/services/reminderScheduler.ts
import { prisma } from "../lib/prisma";
import { notify } from "./notifier";

const DUE_SOON_WINDOW_MS = 24 * 60 * 60 * 1000; // remind 24h before due
const MEETING_REMINDER_MS = 15 * 60 * 1000; // remind 15min before start
const ESCALATION_GRACE_MS = Number(process.env.ESCALATION_GRACE_MS ?? 60 * 60 * 1000); // 1h overdue

async function alreadySent(taskId: string, kind: string) {
  const marker = await prisma.reminder.findUnique({ where: { taskId_kind: { taskId, kind } } });
  return !!marker;
}

async function markSent(taskId: string, kind: string) {
  await prisma.reminder.create({ data: { taskId, kind } }).catch(() => {});
}

/**
 * One scheduler pass. Exported separately from start() so tests can drive it
 * with a fixed clock and grace period.
 */
export async function runReminderPass(now = new Date(), escalationGraceMs = ESCALATION_GRACE_MS) {
  const openWithDue = await prisma.task.findMany({
    where: { status: { not: "COMPLETED" }, dueDate: { not: null } },
    include: {
      assignees: { include: { user: { select: { id: true, name: true, managerId: true } } } },
      assigner: { select: { id: true, name: true } },
    },
  });

  for (const task of openWithDue) {
    const due = task.dueDate!;
    const assigneeIds = task.assignees.map((a) => a.userId);
    const dueLabel = due.toISOString().replace("T", " ").slice(0, 16);

    if (due > now && due.getTime() - now.getTime() <= DUE_SOON_WINDOW_MS) {
      if (!(await alreadySent(task.id, "DUE_SOON"))) {
        await notify(assigneeIds, {
          type: "reminder",
          title: `Task due soon: ${task.title}`,
          body: `Due ${dueLabel}`,
          link: `/tasks?taskId=${task.id}`,
        });
        await markSent(task.id, "DUE_SOON");
      }
    }

    if (due <= now) {
      if (!(await alreadySent(task.id, "OVERDUE"))) {
        await notify([...assigneeIds, task.assignerId], {
          type: "reminder",
          title: `Task overdue: ${task.title}`,
          body: `Was due ${dueLabel}`,
          link: `/tasks?taskId=${task.id}`,
        });
        await markSent(task.id, "OVERDUE");
      }

      // Escalate to each assignee's manager when overdue past the grace period
      // and the assignee never acknowledged the task.
      const overdueFor = now.getTime() - due.getTime();
      if (!task.acknowledgedAt && overdueFor >= escalationGraceMs) {
        if (!(await alreadySent(task.id, "ESCALATION"))) {
          const managerToAssignees = new Map<string, string[]>();
          for (const assignee of task.assignees) {
            const managerId = assignee.user.managerId;
            if (!managerId || assigneeIds.includes(managerId)) continue;
            const names = managerToAssignees.get(managerId) ?? [];
            names.push(assignee.user.name);
            managerToAssignees.set(managerId, names);
          }
          for (const [managerId, names] of managerToAssignees) {
            await notify([managerId], {
              type: "escalation",
              title: `Escalation: unacknowledged overdue task for ${names.join(", ")}`,
              body: `"${task.title}" was due ${dueLabel} and has not been acknowledged.`,
              link: `/tasks?taskId=${task.id}`,
            });
          }
          await markSent(task.id, "ESCALATION");
        }
      }
    }
  }

  // Meeting reminders shortly before start.
  const upcoming = await prisma.meeting.findMany({
    where: {
      reminderSentAt: null,
      startsAt: { gt: now, lte: new Date(now.getTime() + MEETING_REMINDER_MS) },
    },
    include: { attendees: true },
  });
  for (const meeting of upcoming) {
    await notify(
      [...meeting.attendees.map((a) => a.userId), meeting.organizerId],
      {
        type: "meeting_reminder",
        title: `Meeting starting soon: ${meeting.title}`,
        body: `Starts ${meeting.startsAt.toISOString().replace("T", " ").slice(0, 16)}`,
        link: `/meetings`,
      }
    );
    await prisma.meeting.update({
      where: { id: meeting.id },
      data: { reminderSentAt: now },
    });
  }
}

let timer: NodeJS.Timeout | null = null;

export function startReminderScheduler(intervalMs = 60_000) {
  if (timer) return;
  timer = setInterval(() => {
    runReminderPass().catch((err) => console.error("[reminders] pass failed:", err));
  }, intervalMs);
  timer.unref();
  console.log(`[reminders] scheduler running every ${Math.round(intervalMs / 1000)}s`);
}

File: dotnet/TeamCollabWebForms/App_Code/ReminderScheduler.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading;

/// <summary>
/// Background pass every 60s (started from Global.asax):
///  - due-soon reminder 24h before a task's due date
///  - overdue reminder to assignees + assigner
///  - escalation to each assignee's manager when a task is >= 1h overdue and
///    was never acknowledged
///  - meeting reminder 15 minutes before start
/// Reminders table rows act as sent-once markers.
/// </summary>
public static class ReminderScheduler
{
    private static Timer _timer;

    public static void Start()
    {
        if (_timer != null) return;
        _timer = new Timer(_ => SafeRun(), null, TimeSpan.FromSeconds(30), TimeSpan.FromSeconds(60));
    }

    private static void SafeRun()
    {
        try { RunPass(); }
        catch (Exception ex) { System.Diagnostics.Trace.TraceError("Reminder pass failed: " + ex); }
    }

    private static bool AlreadySent(int taskId, string kind)
    {
        return Db.Scalar("SELECT 1 FROM Reminders WHERE TaskId = @t AND Kind = @k",
            Db.P("@t", taskId), Db.P("@k", kind)) != null;
    }

    private static void MarkSent(int taskId, string kind)
    {
        try { Db.Exec("INSERT INTO Reminders (TaskId, Kind) VALUES (@t, @k)", Db.P("@t", taskId), Db.P("@k", kind)); }
        catch { /* unique constraint: already marked by a concurrent pass */ }
    }

    public static void RunPass()
    {
        var tasks = Db.Query(
            @"SELECT t.Id, t.Title, t.DueDate, t.AcknowledgedAt, t.AssignerId
              FROM Tasks t WHERE t.Status <> 'COMPLETED' AND t.DueDate IS NOT NULL");

        foreach (DataRow task in tasks.Rows)
        {
            var taskId = (int)task["Id"];
            var title = (string)task["Title"];
            var due = (DateTime)task["DueDate"];
            var now = DateTime.UtcNow;
            var dueLabel = due.ToString("yyyy-MM-dd HH:mm");

            var assignees = new List<int>();
            var managers = new Dictionary<int, List<string>>(); // managerId -> assignee names
            foreach (DataRow a in Db.Query(
                @"SELECT ta.UserId, u.Name, u.ManagerId
                  FROM TaskAssignees ta JOIN Users u ON u.Id = ta.UserId
                  WHERE ta.TaskId = @t", Db.P("@t", taskId)).Rows)
            {
                assignees.Add((int)a["UserId"]);
                var managerId = a["ManagerId"] as int?;
                if (managerId.HasValue)
                {
                    if (!managers.ContainsKey(managerId.Value)) managers[managerId.Value] = new List<string>();
                    managers[managerId.Value].Add((string)a["Name"]);
                }
            }

            if (due > now && (due - now).TotalHours <= 24 && !AlreadySent(taskId, "DUE_SOON"))
            {
                NotificationService.Notify(assignees, "reminder",
                    "Task due soon: " + title, "Due " + dueLabel, "Tasks.aspx");
                MarkSent(taskId, "DUE_SOON");
            }

            if (due <= now)
            {
                if (!AlreadySent(taskId, "OVERDUE"))
                {
                    NotificationService.Notify(assignees.Concat(new[] { (int)task["AssignerId"] }),
                        "reminder", "Task overdue: " + title, "Was due " + dueLabel, "Tasks.aspx");
                    MarkSent(taskId, "OVERDUE");
                }

                var unacknowledged = task["AcknowledgedAt"] == DBNull.Value;
                if (unacknowledged && (now - due).TotalHours >= 1 && !AlreadySent(taskId, "ESCALATION"))
                {
                    foreach (var pair in managers.Where(p => !assignees.Contains(p.Key)))
                    {
                        NotificationService.Notify(new[] { pair.Key }, "escalation",
                            "Escalation: unacknowledged overdue task for " + string.Join(", ", pair.Value),
                            "\"" + title + "\" was due " + dueLabel + " and has not been acknowledged.",
                            "Tasks.aspx");
                    }
                    MarkSent(taskId, "ESCALATION");
                }
            }
        }

        // Meeting reminders 15 minutes before start.
        var meetings = Db.Query(
            @"SELECT Id, Title, StartsAt, OrganizerId FROM Meetings
              WHERE ReminderSentAt IS NULL
                AND StartsAt > SYSUTCDATETIME()
                AND StartsAt <= DATEADD(MINUTE, 15, SYSUTCDATETIME())");
        foreach (DataRow meeting in meetings.Rows)
        {
            var meetingId = (int)meeting["Id"];
            var attendees = new List<int> { (int)meeting["OrganizerId"] };
            foreach (DataRow a in Db.Query(
                "SELECT UserId FROM MeetingAttendees WHERE MeetingId = @m", Db.P("@m", meetingId)).Rows)
                attendees.Add((int)a["UserId"]);

            NotificationService.Notify(attendees, "meeting_reminder",
                "Meeting starting soon: " + meeting["Title"],
                "Starts " + ((DateTime)meeting["StartsAt"]).ToString("yyyy-MM-dd HH:mm"),
                "Meetings.aspx");
            Db.Exec("UPDATE Meetings SET ReminderSentAt = SYSUTCDATETIME() WHERE Id = @m", Db.P("@m", meetingId));
        }
    }
}
